gadopt/utility.py

```diff
@@ -448,3 +448,26 @@ def wrapper(*args, **kwargs):
         log(f"Time taken for {func.__name__}: {elapsed_time} seconds")
         return result
     return wrapper
+
+
+def absv(u):
+    """Component-wise absolute value of vector for SU stabilisation"""
+    return as_vector([abs(ui) for ui in u])
+
+
+def beta(Pe):
+    """Component-wise beta formula Donea and Huerta (2.47a) for SU stabilisation"""
+    return as_vector([1/tanh(Pei+1e-6) - 1/(Pei+1e-6) for Pei in Pe])
+
+
+def su_nubar(u, J, beta_pe):
+    """SU stabilisation viscosity as a function of velocity, Jaciobian and beta(Pe)"""
+    # SU(PG) ala Donea & Huerta:
+    # Columns of Jacobian J are the vectors that span the quad/hex
+    # which can be seen as unit-vectors scaled with the dx/dy/dz in that direction (assuming physical coordinates x,y,z aligned with local coordinates)
+    # thus u^T J is (dx * u , dy * v)
+    # and following (2.44c) Pe = u^T J / (2*nu)
+    # beta(Pe) is the xibar vector in (2.44a)
+    # then we get artifical viscosity nubar from (2.49)
+
+    return dot(absv(dot(u, J)), beta_pe)/2
```

Comment:
Should `beta()` be an exported function? Would you use it outside of computing $\bar\nu$? If so, maybe it should get a slightly more specific name.

Reply:
Thanks I've moved this inside the su_nubar function